### PR TITLE
iox-#1904 Restore Binding C API compatibility with v2

### LINKED ISSUE #1904: Binding C API compatibility with v2 release is broken
## Required information

The Binding C API compatibility with the v2 release is broken with the change of the log level enum tag `Verbose` to `Trace`.

Compatibility for v3 should be preserved with a deprecation warning and removed in v4.

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] ~~Changelog updated [in the unreleased section][changelog] including API breaking changes~~
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1904

### CODE REVIEW

**Review by mossmaurice — DISMISSED** [review]
LGTM, thanks! Just to make sure, this is already mentioned in the release notes specifically for the C binding, correct?

**Comment by elBoberido** [comment]
> LGTM, thanks! Just to make sure, this is already mentioned in the release notes specifically for the C binding, correct?

yup
 ```
In the C binding the `Iceoryx_LogLevel_Verbose` changed to `Iceoryx_LogLevel_Trace`
```

File: iceoryx_binding_c/include/iceoryx_binding_c/log.h [hunk 1]
```diff
@@ -29,6 +29,14 @@ enum iox_LogLevel
     Iceoryx_LogLevel_Fatal
 };
 
+#ifdef _WIN32
+#define Iceoryx_LogLevel_Verbose Iceoryx_LogLevel_Trace
+#else
+#define Iceoryx_LogLevel_Verbose                                                                                       \
+    _Pragma("GCC warning \"'Iceoryx_LogLevel_Verbose' is deprecated in 3.0 and will be removed in 4.0. Please use "    \
+            "'Iceoryx_LogLevel_Trace'\"") Iceoryx_LogLevel_Trace
```
Comment (dkroenke) [inline iceoryx_binding_c/include/iceoryx_binding_c/log.h:37]: Compiling this with cyclonedds gives the following error:

```
/src/cyclonedds/src/core/ddsi/src/ddsi_shm_transport.c: In function ‘to_iox_loglevel’:
/src/cyclonedds/src/core/ddsi/src/ddsi_shm_transport.c:59:40: error: _Pragma takes a parenthesized string literal
   59 |         case DDSI_SHM_VERBOSE : return Iceoryx_LogLevel_Verbose;
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from src/cyclonedds/src/core/ddsi/src/ddsi_shm_transport.c:13:
/src/core/ddsi/src/ddsi_shm_transport.c:59:40: error: ‘_Pragma’ undeclared (first use in this function)
   59 |         case DDSI_SHM_VERBOSE : return Iceoryx_LogLevel_Verbose;
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~
/src/cyclonedds/src/core/ddsi/src/ddsi_shm_transport.c:59:40: note: each undeclared identifier is reported only once for each function it appears in
/src/cyclonedds/src/core/ddsi/src/ddsi_shm_transport.c:59:40: error: expected ‘;’ before ‘)’ token
   59 |         case DDSI_SHM_VERBOSE : return Iceoryx_LogLevel_Verbose;
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~
src/cyclonedds/src/core/ddsi/src/ddsi_shm_transport.c:59:40: error: expected statement before ‘)’ token
gmake[2]: *** [src/core/CMakeFiles/ddsc.dir/build.make:1210: src/core/CMakeFiles/ddsc.dir/ddsi/src/ddsi_shm_transport.c.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:689: src/core/CMakeFiles/ddsc.dir/all] Error 2
gmake: *** [Makefile:156: all] Error 2
```

I guess the handling with newline and the pragma is a bit tricky.

Shorter version that works:

```suggestion
#define Iceoryx_LogLevel_Verbose                                                                                \
    _Pragma("GCC warning \"Iceoryx_LogLevel_Verbose' is deprecated and will be removed in the next release.\"") \
        Iceoryx_LogLevel_Trace
```

**Review by dkroenke — CHANGES_REQUESTED** [review]
Compile error when compiling with cyclonedds

File: iceoryx_binding_c/include/iceoryx_binding_c/log.h (hunk 1 above)
Comment (elBoberido) [reply iceoryx_binding_c/include/iceoryx_binding_c/log.h]: @dkroenke should be fine now

**Review by mossmaurice — DISMISSED** [review]

**Review by elfenpiff — APPROVED** [review]

**Review by elfenpiff — APPROVED** [review]

File: iceoryx_binding_c/include/iceoryx_binding_c/log.h [hunk 2]
```diff
@@ -29,6 +29,9 @@ enum iox_LogLevel
     Iceoryx_LogLevel_Fatal
 };
 
+/// @deprecated Will be removed with iceoryx 4.0. This typedf is required for Cyclone DDS.
```
Comment (mossmaurice) [inline iceoryx_binding_c/include/iceoryx_binding_c/log.h:32]:
```suggestion
/// @deprecated Will be removed with iceoryx 4.0. This typedef is required for Cyclone DDS.
```

**Review by mossmaurice — CHANGES_REQUESTED** [review]

File: iceoryx_binding_c/include/iceoryx_binding_c/log.h (hunk 2 above)
Comment (elBoberido) [reply iceoryx_binding_c/include/iceoryx_binding_c/log.h]: :sweat_smile:

**Review by mossmaurice — APPROVED** [review]

**Review by dkroenke — APPROVED** [review]
Tested on latest cyclonedds master and releases/0.9.x branch, looks good.

**Pull request merged**